printer.go

```diff
@@ -197,35 +199,44 @@ func (p *printer) printStruct() {
 		}
 	}
 
-	if p.value.NumField() == 0 {
+	var fields []int
+	for i := 0; i < p.value.NumField(); i++ {
+		field := p.value.Type().Field(i)
+		value := p.value.Field(i)
+		// ignore unexported if needed
+		if p.exportedOnly && field.PkgPath != "" {
+			continue
+		}
+		// ignore fields if zero value, or explicitly set
+		if tag := field.Tag.Get("pp"); tag != "" {
+			parts := strings.Split(tag, ",")
+			if len(parts) == 2 && parts[1] == "omitempty" && valueIsZero(value) {
+				continue
+			}
+			if parts[0] == "-" {
+				continue
+			}
+		}
+		fields = append(fields, i)
+	}
+
+	if len(fields) == 0 {
 		p.print(p.typeString() + "{}")
 		return
 	}
 
 	p.println(p.typeString() + "{")
 	p.indented(func() {
-		for i := 0; i < p.value.NumField(); i++ {
+		for _, i := range fields {
 			field := p.value.Type().Field(i)
 			value := p.value.Field(i)
```

Comment:
Does this still work correctly even if some fields are skipped in the middle?

Reply:
Yep, the tests still pass (iirc in which the first field is skipped)

Reply:
ah, sorry, you maintained a slice of indices. That should work, of course. Never mind.